Fixes SPARK-12910: R version for installing sparkR

napsternxg opened the pull request:
Testing code:

```
$ ./install-dev.sh 
USING R_HOME = /usr/bin
ERROR: this R is version 2.15.1, package 'SparkR' requires R >= 3.0
```

Using the new argument:

```
$ ./install-dev.sh /content/username/SOFTWARE/R-3.2.3
USING R_HOME = /content/username/SOFTWARE/R-3.2.3/bin
* installing *source* package â€˜SparkRâ€™ ...
** R
** inst
** preparing package for lazy loading
Creating a new generic function for â€˜colnamesâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜colnames<-â€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜covâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜na.omitâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜filterâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜intersectâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜sampleâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜transformâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜subsetâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜summaryâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜lagâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜rankâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜sdâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜varâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜predictâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜rbindâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜lapplyâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜Filterâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜aliasâ€™ from package â€˜statsâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜substrâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜%in%â€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜meanâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜uniqueâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜nrowâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜ncolâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜headâ€™ from package â€˜utilsâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜factorialâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜atan2â€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜ifelseâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
** help
No man pages found in package  â€˜SparkRâ€™ 
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (SparkR)

```

napsternxg committed:
Fixes SPARK-12910: R version for installing sparkR

I have added an option for passing the R_HOME path to the
`install-dev.sh` file which fixes SPARK-12910.

The code works and has been tested.

napsternxg committed:
Merge branch 'master' of github.com:apache/spark

AmplabJenkins commented:
Can one of the admins verify this patch?

napsternxg committed:
Merge branch 'master' of github.com:apache/spark

shivaram commented:
Jenkins, ok to test

shivaram commented on R/install-dev.sh:
```diff
@@ -35,12 +35,19 @@ LIB_DIR="$FWDIR/lib"
+  then
+    R_HOME="$1/bin"
+   else
+    R_HOME="$(dirname $(which R))"
```
Can we just use a `R_HOME` if it is set instead of using a command line argument ?

napsternxg replied:
In R there there is usually no `R_HOME` like there is in python so this might confuse users.  We should use the default R when the variable is not passed. 

Or I can use a variable like `SPARK_R_HOME` which specifies the R home directory ?

napsternxg replied:
Also if using an environment variable then the code for running sparkR would also have to be edited to use the R version from this path.

shivaram replied:
`R_HOME` is a reasonably well defined concept ? http://www.inside-r.org/r-doc/base/R.home 

Regardless I think you will need to use a different R binary for running sparkR if you don't want it to use the default R ?

napsternxg replied:
I have made the fix. 

Here are my new test results:

```
$ unset R_HOME
ram:/content/username/SOFTWARE/spark/R $ ./install-dev.sh 
USING R_HOME = 
ERROR: this R is version 2.15.1, package 'SparkR' requires R >= 3.0
* removing â€˜/content/username/SOFTWARE/spark/R/lib/SparkRâ€™
ram:/content/username/SOFTWARE/spark/R $ export R_HOME=/content/username/SOFTWARE/R-3.2.3
ram:/content/username/SOFTWARE/spark/R $ ./install-dev.sh 
USING R_HOME = /content/username/SOFTWARE/R-3.2.3
* installing *source* package â€˜SparkRâ€™ ...
** R
** inst
** preparing package for lazy loading
Creating a new generic function for â€˜colnamesâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜colnames<-â€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜covâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜na.omitâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜filterâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜intersectâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜sampleâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜transformâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜subsetâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜summaryâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜lagâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜rankâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜sdâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜varâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜predictâ€™ in package â€˜SparkRâ€™
Creating a new generic function for â€˜rbindâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜lapplyâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜Filterâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜aliasâ€™ from package â€˜statsâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜substrâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜%in%â€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜meanâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜uniqueâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜nrowâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜ncolâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜headâ€™ from package â€˜utilsâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜factorialâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜atan2â€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
Creating a generic function for â€˜ifelseâ€™ from package â€˜baseâ€™ in package â€˜SparkRâ€™
** help
No man pages found in package  â€˜SparkRâ€™ 
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (SparkR)

```

napsternxg replied:
Updated my pull request.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49709/
Test FAILed.

napsternxg committed:
Fixes SPARK-12910: R version for installing sparkR

Added the option for using the environment variable R_HOME which is used
by `install-dev.sh` script for creating sparkR libraries. Fixes SPARK-12910.

The code works and has been tested.

napsternxg committed:
Merge branch 'master' of github.com:apache/spark

SparkQA commented:
**[Test build #49721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49721/consoleFull)** for PR 10836 at commit [`0b3960a`](https://github.com/apache/spark/commit/0b3960a29dd47ae12c9534806191cfb098139888).

SparkQA commented:
**[Test build #49721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49721/consoleFull)** for PR 10836 at commit [`0b3960a`](https://github.com/apache/spark/commit/0b3960a29dd47ae12c9534806191cfb098139888).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49721/
Test PASSed.

shivaram commented:
LGTM. cc @felixcheung

felixcheung commented on R/README.md:
```diff
@@ -1,6 +1,16 @@
 SparkR is an R package that provides a light-weight frontend to use Spark from R.
+### Installing sparkR
+
+Libraries of sparkR need to be created in `$SPARK_HOME/R/lib`. This can be done by running the script `$SPARK_HOME/R/install-dev.sh`.
```
If one were running from R studio with the steps in `Using SparkR from RStudio` below he wouldn't have to install or run install-dev.sh though - could we clarify that?

napsternxg replied:
I am running this on a machine with no X-server hence no R-studio, so this kind of functionality will be needed for users like me. 

Even for R-studio I feel sparkR needs to be compiled with the version of R >= 3.0

felixcheung replied:
To clarify, I'm referring the part about "sparkR need to be created in `$SPARK_HOME/R/lib`. This can be done by running the script `$SPARK_HOME/R/install-dev.sh`."

napsternxg replied:
I think this is needed as mentioned in running sparkR from R studio. The
script there tries to access the lib location which might not be present by
default in the sparkR folder if the wrong version of R is selected by
default.


*Regards,*
*Shubhanshu Mishra*
Research Assistant,
iSchool at University of Illinois at Urbana-Champaign
--------------------------------------------------
*Website:* http://shubhanshu.com
*LinkedIn Profile: *http://www.linkedin.com/in/shubhanshumishra

Blog <http://smexyyweby.wordpress.com/>  || Facebook
<http://www.facebook.com/shubhanshu.mishra>  ||  Twitter
<http://www.twitter.com/TheShubhanshu>  || LinkedIn
<http://www.linkedin.com/in/shubhanshumishra>

On Tue, Jan 19, 2016 at 10:03 PM, Felix Cheung <notifications@github.com>
wrote:

> In R/README.md
> <https://github.com/apache/spark/pull/10836#discussion_r50212459>:
>
> > @@ -1,6 +1,16 @@
> >  # R on Spark
> >
> >  SparkR is an R package that provides a light-weight frontend to use Spark from R.
> > +### Installing sparkR
> > +
> > +Libraries of sparkR need to be created in `$SPARK_HOME/R/lib`. This can be done by running the script `$SPARK_HOME/R/install-dev.sh`.
>
> To clarify, I'm referring the part about "sparkR need to be created in
> $SPARK_HOME/R/lib. This can be done by running the script
> $SPARK_HOME/R/install-dev.sh."
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/10836/files#r50212459>.
>

napsternxg replied:
I think this is needed as mentioned in running sparkR from R studio. The script there tries to access the lib location which might not be present by default in the sparkR folder if the wrong version of R is selected by default.

felixcheung replied:
actually, this line
```.libPaths(c(file.path(Sys.getenv("SPARK_HOME"), "R", "lib"), .libPaths()))```
added SPARK_HOME/R/lib into R's lib path and allows R, any running version, to load SparkR package from there - SparkR packages does not need to be installed with `R CMD INSTALL` (in install-dev.sh) at all.

napsternxg replied:
I am sorry if my I was not clear before but what I mean is the following:
1. In order for the code `.libPaths(c(file.path(Sys.getenv("SPARK_HOME"), "R", "lib"), .libPaths()))` to work, the directory `$SPARK_HOME/R/lib` needs to exist. When we build spark using `build/mvn -DskipTests clean package` this directory is not created by default. Hence we have to run `install-dev.sh` in order to use SparkR from an R shell. 

Also, if we look at the code in `install-dev.sh`, the following lines actually create the `lib` directory.
```
FWDIR="$(cd `dirname $0`; pwd)"
LIB_DIR="$FWDIR/lib"

mkdir -p $LIB_DIR
```

felixcheung replied:
ok I think I get your point now.
I guess we are saying this README.md is more for developer, so I'm ok with what you have here.
There are users that are not building Spark from source and are running with the binary release, in which case the SPARK_HOME/R/lib is there and they would not need to install the SparkR package. Similarly when running SparkR with a cluster manager, on the worker nodes SparkR would not need to be installed either. I agree they are possibly outside the scope of this file.

napsternxg replied:
Great. So is this PR ready for a merge into the master ?

felixcheung commented:
looks good.

felixcheung commented:
looks good.

shivaram commented:
Merging this to master
